872 merge name into patient data

clemens-msupply opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

Fixes #872

# 👩🏻‍💻 What does this PR do? 
Adds a `documentDraft` node to the PatientNode. This `documentDraft` contains the document data including changes from the name_row (if there are any). If there are no change in the name_row the `documentDraft` is the same as `document.data`

On the frontend this `documentDraft` is used for patient editing. When opening a remotely edited patient these changes now show up in omSupply. When a user saves a patient the remote name_row changes get merged back into the document.

This solution was quite easy to implement. However, the disadvantage is that it is not 100% correct in recording what a user did. For example, when looking at the document edit history it will look like as if the user made the mSupply to the patient document, i.e. the mSupply change and the omSupply change are squashed into one document version. This could, for example, be fixed by adding a document version only containing the name row changes. I don't think this is a problem at the moment though, and I kept it simple....

On the frontend side I did some refactoring including adding some missing dependency array entries and the fixing the resulting rendering bugs.

# 🧪 How has/should this change been tested? 
- Add it a patient in omSupply
- Then edit the same patient in mSupply, e.g. change address or first name
- After a sync the mSupply change should show up in omSupply

clemens-msupply commented on client/packages/common/src/hooks/useDialog/useDialog.tsx:
```diff
@@ -103,8 +103,8 @@ export const useDialog = (dialogProps?: DialogProps): DialogState => {
-  const showDialog = () => setOpen(true);
-  const hideDialog = () => setOpen(false);
+  const showDialog = useCallback(() => setOpen(true), []);
+  const hideDialog = useCallback(() => setOpen(false), []);
```
These methods were changing in every render...

clemens-msupply commented on server/service/src/programs/patient/patient_updated.rs:
```diff
@@ -43,6 +43,25 @@ pub fn update_patient_row(
+    update_timestamp: &DateTime<Utc>,
+    patient: SchemaPatient,
+    existing_name: Option<NameRow>,
+) -> Result<NameRow, UpdateProgramPatientError> {
```
I refactor this to be able to use it in the new test.

clemens-msupply commented on server/service/src/programs/patient/patient_updated.rs:
```diff
@@ -101,22 +118,26 @@ pub fn update_patient_row(
+            SchemaGender::TransgenderFemaleHormone => Gender::TransgenderFemaleHormone,
+            SchemaGender::TransgenderFemaleSurgical => Gender::TransgenderFemaleSurgical,
+            SchemaGender::Unknown => Gender::Unknown,
+            SchemaGender::NonBinary => Gender::NonBinary,
```
The schema was missing some gender types that are in omSupply...

andreievg reviewed (commented):
Thanks, in general I think I understand the functionality and in most part it makes sense.

Had a few question on front end thought. Mainly about watching functions as dependencies ? And questions about refactor. 

Can discuss these in person, might be easier.

Also another question:

https://github.com/msupply-foundation/open-msupply/blob/3eff842119dfb047ee454a26159239b04160d9f9/client/packages/system/src/Patient/PatientView/PatientView.tsx#L228-L232

Should we also wait for:

https://github.com/msupply-foundation/open-msupply/blob/3eff842119dfb047ee454a26159239b04160d9f9/client/packages/system/src/Patient/PatientView/PatientView.tsx#L99

And how do we differentiate between waiting for patient to be fetched when producing inputData and default undefined currentPatient (before it's either fetched or it's fetched and doesn't exist)

clemens-msupply commented on client/packages/system/src/Patient/CreatePatientModal/CreatePatientModal.tsx:
```diff
@@ -32,15 +31,6 @@ interface CreatePatientModal {
-
 export const CreatePatientModal: FC<CreatePatientModal> = ({ onClose }) => {
   const { data: documentRegistryResponse, isLoading } =
     useDocumentRegistry.get.documentRegistries({
```
Thanks the category version is IMHO the better solution (both work) so I changed it accordingly.

andreievg commented:
I had thought about this (sorry if you've already mentioned this). Previously we used the latest document in sync to update the patient. Now we would be using the latest state of name record to add on top of latest document. This would mean if two remote omSupply sites are editing patients, through documents, the "latest document" would not be respected ?

clemens-msupply commented on server/service/src/programs/patient/patient_updated.rs:
```diff
@@ -43,6 +43,25 @@ pub fn update_patient_row(
+    let existing_name = name_repo.find_one_by_id(&patient.id)?;
+
+    let name_upsert = patient_to_name_row(store_id, update_timestamp, patient, existing_name)?;
+    if is_sync_update {
```
No! this method was only called with `false`. I simplified it, thanks.

clemens-msupply commented on server/service/src/programs/patient/patient_updated.rs:
```diff
@@ -128,15 +149,79 @@ pub fn update_patient_row(
+///
+/// The patient can divert from the document data when, for example, the patient details have been
+/// changed in mSupply.
+pub fn patient_draft_document(patient: &Patient, document_data: SchemaPatient) -> SchemaPatient {
```
Yes good idea

clemens-msupply commented:
> I had thought about this (sorry if you've already mentioned this). Previously we used the latest document in sync to update the patient. Now we would be using the latest state of name record to add on top of latest document. This would mean if two remote omSupply sites are editing patients, through documents, the "latest document" would not be respected ?

When there were two edits to the patient's document the two versions would come together after a sync. During sync only the latest document would get translated to the `name` table. The draft document data would then be the latest document data (since the draft is only calculated when there is no document data).

clemens-msupply commented on server/service/src/programs/patient/patient_updated.rs:
```diff
@@ -128,15 +149,79 @@ pub fn update_patient_row(
+        date_of_birth_is_estimated: document_data.date_of_birth_is_estimated,
+        is_deceased: Some(patient.is_deceased || document_data.is_deceased.unwrap_or(false)),
+        notes: document_data.notes,
+        passport_number: document_data.passport_number,
```
thanks done

clemens-msupply commented on server/graphql/types/src/types/program/patient.rs:
```diff
@@ -116,6 +117,33 @@ impl PatientNode {
+
+        let document_data = serde_json::from_value(document_data)
+            .map_err(|e| StandardGraphqlError::InternalError(format!("{}", e)).extend())?;
+        let draft = patient_draft_document(&self.patient, document_data);
```
thanks done

clemens-msupply commented:
> https://github.com/msupply-foundation/open-msupply/blob/3eff842119dfb047ee454a26159239b04160d9f9/client/packages/system/src/Patient/PatientView/PatientView.tsx#L99
> 
> And how do we differentiate between waiting for patient to be fetched when producing inputData and default undefined currentPatient (before it's either fetched or it's fetched and doesn't exist)

I added the isLoading check and return a spinner if still loading. Think this will also handle the outlined problem (unsure if it actual was a problem).

The comments I left uncommented are all related to the react lint changes. These changes where mainly to make the linter happy...

Additionally, I also fixed/changed one sync test. The test failed because the name table don't get updated durring sync when integrating documents anymore. I changed the test to use enrolments instead where the test case is still valid.

Thanks for the review and sorry for the late reply! I think I addressed all comments.

clemens-msupply commented on server/service/src/programs/patient/patient_updated.rs:
```diff
@@ -128,15 +149,79 @@ pub fn update_patient_row(
+        birth_place: document_data.birth_place,
+        marital_status: document_data.marital_status,
+        contacts: document_data.contacts,
+        extension: document_data.extension,
```
There are currently:
1) name row
2) common patient document data, which contains extra data not in the name row
3) patient document extension for country specific data not in the common patient document data

For the patient, the enrolment and the encounter data only the common fields that needs to be queried are in a table (but they all have additional common fields).

Think currently it wouldn't cause much problem moving the new common data fields from 2) to 3) so that the main patient document data can be mapped 1 to 1 to the name row. If needed we can then extent name_row and common patient document data type.

Main problem is that its used in the schemas and reports which would need to be updated and tested.

(Another thing to consider but which hasn't really be done is to have a FHIR friendly patient data structure.)
